modules/dag-l1/src/main/scala/org/tessellation/dag/l1/StateChannel.scala

```diff
@@ -0,0 +1,173 @@
+import org.tessellation.dag.domain.block.DAGBlock
+import org.tessellation.dag.l1.domain.block.{BlockService, BlockStorage}
+import org.tessellation.dag.l1.domain.consensus.block.BlockConsensusCell.isReadyForBlockConsensus
+import org.tessellation.dag.l1.domain.consensus.block.BlockConsensusInput.{
+  OwnRoundTrigger,
+  OwnerBlockConsensusInput,
+  PeerBlockConsensusInput
+}
+import org.tessellation.dag.l1.domain.consensus.block.BlockConsensusOutput.FinalBlock
+import org.tessellation.dag.l1.domain.consensus.block.config.ConsensusConfig
+import org.tessellation.dag.l1.domain.consensus.block.http.p2p.clients.BlockConsensusClient
+import org.tessellation.dag.l1.domain.consensus.block.storage.ConsensusStorage
+import org.tessellation.dag.l1.domain.consensus.block.{BlockConsensusCell, BlockConsensusContext, BlockConsensusInput}
+import org.tessellation.dag.l1.domain.transaction.TransactionStorage
+import org.tessellation.kernel.Cell.NullTerminal
+import org.tessellation.kernel.{CellError, Ω}
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.domain.cluster.storage.ClusterStorage
+import org.tessellation.sdk.domain.gossip.Gossip
+import org.tessellation.sdk.domain.node.NodeStorage
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.Signed
+
+import eu.timepit.refined.auto.autoUnwrap
+import fs2.{Pipe, Stream}
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+class StateChannel[F[_]: Async: KryoSerializer: SecurityProvider: Random](
+  blockConsensusClient: BlockConsensusClient[F],
+  blockService: BlockService[F],
+  blockStorage: BlockStorage[F],
+  blockValidator: BlockValidator[F],
+  clusterStorage: ClusterStorage[F],
+  consensusConfig: ConsensusConfig,
+  consensusStorage: ConsensusStorage[F],
+  gossip: Gossip[F],
+  keyPair: KeyPair,
+  myId: PeerId,
+  nodeStorage: NodeStorage[F],
+  peerBlockConsensusInputQueue: Queue[F, Signed[PeerBlockConsensusInput]],
+  peerBlockQueue: Queue[F, Signed[DAGBlock]],
+  transactionStorage: TransactionStorage[F]
+) {
+
+  private implicit val logger = Slf4jLogger.getLogger[F]
+
+  private val blockConsensusContext =
+    BlockConsensusContext[F](
+      blockConsensusClient,
+      blockStorage,
+      blockValidator,
+      clusterStorage,
+      consensusConfig,
+      consensusStorage,
+      keyPair,
+      myId,
+      transactionStorage
+    )
+
+  private def enoughPeersForConsensus: F[Boolean] =
+    clusterStorage.getPeers
+      .map(_.filter(p => isReadyForBlockConsensus(p.state)))
+      .map(_.size >= consensusConfig.peersCount)
+
+  private val ownerBlockConsensusInputs: Stream[F, OwnerBlockConsensusInput] = Stream
+    .awakeEvery(5.seconds)
+    .evalFilter(_ => consensusStorage.ownConsensus.get.map(_.isEmpty))
+    .evalFilter(_ => nodeStorage.getNodeState.map(isReadyForBlockConsensus))
+    .evalFilter(_ => enoughPeersForConsensus)
```

Comment:
what about:

```scala
private def canStartOwnConsensus: F[Boolean] =
  for { 
    ownRoundInProgress <- ...
    stateReadyForConsensus <- ...
    enoughPeers <- ...
   } yield !ownRoundInProgress && stateReadyForConsensus && enoughPeers
   
Stream.awakeEvery(5.seconds)
   .evalFilter(_ => canStartOwnConsensus)
   .as(OwnRoundTrigger)
```
?

Reply:
I guess the multiple `evalFilter` solution may be more performant. If the first condition fails the next ones are not even evaluated. Although that might be micro-optimizing.